modules/dex/libdexc/adapter.go

```diff
@@ -0,0 +1,266 @@
+// CoreAdapter manages a Core instance and possibly a Server instance for
+// a library user.
+type CoreAdapter struct {
+	ctx           context.Context
+	kill          context.CancelFunc
+	inited        uint32
+	logLevel      slog.Level
+	serverRunning uint32
+	core          *core.Core
+	webServer     *dex.ConnectionMaster
+
+	preInitMethods map[string]callHandler
+	directMethods  map[string]callHandler
+}
+
+func NewCoreAdapter() *CoreAdapter {
+	c := &CoreAdapter{}
+
+	c.preInitMethods = map[string]callHandler{
+		"IsInitialized": c.isInitialized,
+	}
+
+	c.directMethods = map[string]callHandler{
+		// Some control functions.
+		"startServer": c.startServer,
+		"shutdown":    c.shutdown,
+		// Pass-throughs to Core
+		"Init":         c.init,
+		"CreateWallet": c.createWallet,
+		"UpdateWallet": c.updateWallet,
+		"User":         c.user,
+		"Register":     c.register,
+		"Login":        c.login,
+		"Logout":       c.logout,
+		"DexConfig":    c.getDexConfig,
+	}
+
+	return c
+}
+
+func (c *CoreAdapter) startCore(raw json.RawMessage) error {
+	if !atomic.CompareAndSwapUint32(&c.inited, 0, 1) {
+		return fmt.Errorf("already initialized")
+	}
+
+	form := new(struct {
+		DBPath   string      `json:"dbPath"`
+		Net      dex.Network `json:"net"`
+		LogLevel slog.Level  `json:"logLevel"`
+	})
+	if err := json.Unmarshal(raw, form); err != nil {
+		return err
+	}
+
+	err := os.MkdirAll(filepath.Dir(form.DBPath), 0777)
+	if err != nil {
+		return err
+	}
+
+	c.ctx, c.kill = context.WithCancel(context.Background())
+	ccore, err := core.New(&core.Config{
+		DBPath: form.DBPath,
+		Net:    form.Net,
+		Logger: dex.StdOutLogger("CORE", form.LogLevel),
+	})
+	if err != nil {
+		return fmt.Errorf("error creating client core: %v", err)
+	}
+	c.core = ccore
+	c.logLevel = form.LogLevel
+
+	go ccore.Run(c.ctx)
+	<-ccore.Ready()
+
+	return nil
+}
+
+func (c *CoreAdapter) startServer(raw json.RawMessage) (string, error) {
+	if !atomic.CompareAndSwapUint32(&c.serverRunning, 0, 1) {
+		return "", fmt.Errorf("already initialized")
+	}
+	form := new(struct {
+		WebAddr string `json:"webaddr"`
+		SiteDir string `json:"sitedir"`
+	})
+	if err := json.Unmarshal(raw, form); err != nil {
+		return "", err
+	}
+	webSrv, err := webserver.New(c.core, form.WebAddr, form.SiteDir, dex.StdOutLogger("SRVR", c.logLevel), false, false)
+	if err != nil {
+		return "", fmt.Errorf("Error creating web server: %v", err)
+	}
+	cm := dex.NewConnectionMaster(webSrv)
+	err = cm.Connect(c.ctx)
+	if err != nil {
+		return "", fmt.Errorf("Error starting web server: %v", err)
+	}
+	c.webServer = cm
+	go func() {
+		cm.Wait()
+		atomic.StoreUint32(&c.serverRunning, 0)
+	}()
+	return "", nil
+}
+
+func (c *CoreAdapter) shutdown(json.RawMessage) (string, error) {
+	if !atomic.CompareAndSwapUint32(&c.inited, 1, 0) || c.kill == nil {
+		return "", fmt.Errorf("already shut down")
+	}
+	if !atomic.CompareAndSwapUint32(&c.serverRunning, 1, 0) {
+		return "", fmt.Errorf("server not running initialized")
```

Comment:
I'm not sure why the webserver's status should prevent stopping Core.  I'm actually wondering why `shutdown` can't just run `kill` regardless of both subsystems' states.  It's a context.CancelFunc